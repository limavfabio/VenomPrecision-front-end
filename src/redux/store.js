import { configureStore } from '@reduxjs/toolkit';
import productsReducer from './productsSlice';
import reservationReducer from './reservationSlice';
import userReducer from './userSlice';

const store = configureStore({
  reducer: {
    products: productsReducer,
    reservations: reservationReducer,
    user: userReducer,
  },
});

export default store;
