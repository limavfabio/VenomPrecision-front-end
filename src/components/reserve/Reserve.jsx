import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, useParams, useLocation } from 'react-router-dom';
import { postReservation } from '../../redux/reservationSlice';
import ReserveCalender from './ReserveCalender';
import ReserveCity from './ReserveCity';
import { fetchProductWithId } from '../../redux/productsSlice';

const Reserve = () => {
  const dispatch = useDispatch();
  const redirect = useNavigate();

  // Fetch user from redux store
  const user = useSelector((state) => state.user);

  // Fetch product based on the parameter id
  const { productId } = useParams();
  const product = useSelector((state) => state.products.value);
  useEffect(() => {
    dispatch(fetchProductWithId(productId));
  }, [dispatch]);

  // Redirect to login page if user is not logged in
  useEffect(() => {
    if (!user.id) {
      redirect('/login');
    }
  }, [user, redirect]);

  const [date, setDate] = useState(null);
  const [city, setCity] = useState(null);
  const history = useLocation();
  const { data } = history.state;
  const userId = useSelector(state => state.user.id)

  const navigate = useNavigate();

  const handleCalender = (date) => {
    setDate(date);
  };
  const handleCity = (city) => {
    setCity(city);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const postData = {
      date,
      user_id: parseInt(userId),
      product_id: data.id,
      city:city
    };

    // Create a new reservation
    dispatch(postReservation({ postData }));
    navigate(`/reservations`, { state: { userId, data } });
  };

  const bgImg = {
    backgroundImage: `url(${product.image})`,
    backgroundSize: 'cover',
    backgroundPosition: 'center',
    backgroundRepeat: 'no-repeat',
  };

  return (
    <div style={bgImg}>
      <div className="flex h-screen w-full flex-col items-center justify-center bg-[#96BF01] bg-opacity-90 text-center text-white">
        <h2 className="text-2xl font-bold uppercase tracking-[.6rem]">
          book a vespa test-drive
        </h2>
        <hr className="my-5 w-2/3 border-[#A1C839]" />
        <p className="w-1/2 text-xs leading-6">
          There are 34 different versions of the Vespa. Today five series are in
          production: the
        </p>

        <div className="mt-5 flex flex-col items-center justify-center gap-5 md:flex-row">
          <ReserveCalender handleCalender={handleCalender} />
          <ReserveCity handleCity={handleCity}/>

          <button
            type="button"
            className="rounded-full border-2  border-white bg-white px-8 py-2 text-xs text-[#97BF0F] transition delay-150 duration-200 ease-in-out  hover:-translate-y-1 hover:scale-110  hover:bg-[#97BF0F] hover:text-white"
            onClick={handleSubmit}
          >
            Book Now
          </button>
        </div>
      </div>
    </div>
  );
};

export default Reserve;
