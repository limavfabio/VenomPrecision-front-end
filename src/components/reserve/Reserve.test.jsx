import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, useLocation } from 'react-router-dom';
import store from '../../redux/store';
import Reserve from './Reserve';
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: jest.fn(),
  }));
describe('Main Home', () => {
  it('expects to render a snapshot of the Home component', () => {
    const mockLocation = {
        state: { data: 1 },
      };
      useLocation.mockReturnValue(mockLocation);
  
    const episodeTest = render(
      <Provider store={store}>
        <BrowserRouter>
          <Reserve/>
        </BrowserRouter>
      </Provider>,
    );
    expect(episodeTest).toMatchSnapshot();
  });
});