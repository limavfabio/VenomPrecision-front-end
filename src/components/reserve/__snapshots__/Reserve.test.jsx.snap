// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Home expects to render a snapshot of the Home component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="background-image: url(); background-size: cover; background-position: center; background-repeat: no-repeat;"
      >
        <div
          class="flex h-screen w-full flex-col items-center justify-center bg-[#96BF01] bg-opacity-90 text-center text-white"
        >
          <h2
            class="text-2xl font-bold uppercase tracking-[.6rem]"
          >
            book a Awesome Weapon
          </h2>
          <hr
            class="my-5 w-2/3 border-[#A1C839]"
          />
          <p
            class="w-1/2 text-xs leading-6"
          >
            We have a wide range of weapons and accessories for you to choose from. You can book your product for as long as you wish.
          </p>
          <div
            class="mt-5 flex flex-col items-center justify-center gap-5 md:flex-row"
          >
            <div>
              <button
                class="flex items-center justify-between gap-3 rounded-full border-2 border-[#BED86B] bg-[#97BF0F] px-4 py-2 text-xs"
                type="button"
              >
                <input
                  class=" appearance-none border-none bg-transparent text-black outline-none"
                  min="2023-06-23"
                  style="color: white; background-color: transparent; appearance: none; outline: none;"
                  type="date"
                  value="2023-06-23"
                />
              </button>
            </div>
            <div
              class=" inline-block "
            >
              <button
                class="flex items-center justify-between gap-2 rounded-full border-2 border-[#BED86B] bg-[#97BF0F] px-6 py-2 text-xs"
                type="button"
              >
                <p>
                  New York
                </p>
                <svg
                  class="rotate-90 text-xl"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10.804 8 5 4.633v6.734L10.804 8zm.792-.696a.802.802 0 0 1 0 1.392l-6.363 3.692C4.713 12.69 4 12.345 4 11.692V4.308c0-.653.713-.998 1.233-.696l6.363 3.692z"
                  />
                </svg>
              </button>
            </div>
            <button
              class="rounded-full border-2  border-white bg-white px-8 py-2 text-xs text-[#97BF0F] transition delay-150 duration-200 ease-in-out  hover:-translate-y-1 hover:scale-110  hover:bg-[#97BF0F] hover:text-white"
              type="button"
            >
              Book Now
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="background-image: url(); background-size: cover; background-position: center; background-repeat: no-repeat;"
    >
      <div
        class="flex h-screen w-full flex-col items-center justify-center bg-[#96BF01] bg-opacity-90 text-center text-white"
      >
        <h2
          class="text-2xl font-bold uppercase tracking-[.6rem]"
        >
          book a Awesome Weapon
        </h2>
        <hr
          class="my-5 w-2/3 border-[#A1C839]"
        />
        <p
          class="w-1/2 text-xs leading-6"
        >
          We have a wide range of weapons and accessories for you to choose from. You can book your product for as long as you wish.
        </p>
        <div
          class="mt-5 flex flex-col items-center justify-center gap-5 md:flex-row"
        >
          <div>
            <button
              class="flex items-center justify-between gap-3 rounded-full border-2 border-[#BED86B] bg-[#97BF0F] px-4 py-2 text-xs"
              type="button"
            >
              <input
                class=" appearance-none border-none bg-transparent text-black outline-none"
                min="2023-06-23"
                style="color: white; background-color: transparent; appearance: none; outline: none;"
                type="date"
                value="2023-06-23"
              />
            </button>
          </div>
          <div
            class=" inline-block "
          >
            <button
              class="flex items-center justify-between gap-2 rounded-full border-2 border-[#BED86B] bg-[#97BF0F] px-6 py-2 text-xs"
              type="button"
            >
              <p>
                New York
              </p>
              <svg
                class="rotate-90 text-xl"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10.804 8 5 4.633v6.734L10.804 8zm.792-.696a.802.802 0 0 1 0 1.392l-6.363 3.692C4.713 12.69 4 12.345 4 11.692V4.308c0-.653.713-.998 1.233-.696l6.363 3.692z"
                />
              </svg>
            </button>
          </div>
          <button
            class="rounded-full border-2  border-white bg-white px-8 py-2 text-xs text-[#97BF0F] transition delay-150 duration-200 ease-in-out  hover:-translate-y-1 hover:scale-110  hover:bg-[#97BF0F] hover:text-white"
            type="button"
          >
            Book Now
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
